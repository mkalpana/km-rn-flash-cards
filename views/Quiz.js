import React, { Component } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import { Button } from '../components';
import { fetchDeck } from '../actions';
import { connect } from 'react-redux';
import { blue, black, green, red } from '../utils/colors';
import { get } from 'lodash';
import { clearLocalNotification, setLocalNotification } from '../utils/notification';

const INITIAL_STATE = {
  currentQuestionIndex: 0,
  correctAnswers: 0,
  showQuestion: true
};

class Quiz extends Component {
  state = INITIAL_STATE;

  componentDidMount() {
    const { navigation } = this.props;
    clearLocalNotification()
      .then(setLocalNotification);
    const title = get(navigation, 'state.params.title', '');
    if (title) {
      this.props.fetchDeck(title);
    }
  }

  _onAnswer = (correct) => {
    this.setState(state => ({
      showQuestion: true,
      correctAnswers: correct ? state.correctAnswers + 1 : state.correctAnswers,
      currentQuestionIndex: state.currentQuestionIndex + 1,
    }));
  };

  resetQuiz = () => {
    this.setState(INITIAL_STATE);
  };

  renderQuestion = (question, index) => {
    const { deck } = this.props;
    const { showQuestion, currentQuestionIndex } = this.state;
    const showCurrentQuestion = currentQuestionIndex <= deck.questions.length - 1 && currentQuestionIndex  === index;
    return ( showCurrentQuestion &&
      <View key={index} style={styles.container}>
        <Text style={styles.questionNumber}>{index+1} / {deck.questions.length} </Text>
        <View style={styles.textContainer}>
          { showQuestion ? (
            <View>
              <Text style={styles.textStyle}>{question.question}</Text>
              <Text style={styles.textBtnStyle} onPress={() => this.setState({ showQuestion: false })}>
                Answer
              </Text>
            </View>
          ): (
            <View>
              <Text style={styles.textStyle}>{question.answer}</Text>
              <Text style={styles.textBtnStyle} onPress={() => this.setState({ showQuestion: true })}>
                Question
              </Text>
            </View>
          )}
        </View>
      <View style={styles.inputContainer}>
        <Button
          style={styles.btnStyle}
          buttonStyle={{ backgroundColor: green, borderWidth: 0 }}
          label="Correct"
          onPress={() => this._onAnswer(true)}
        />
        <Button
          style={styles.btnStyle}
          buttonStyle={{ backgroundColor: red, borderWidth: 0 }}
          label="Incorrect"
          onPress={() => this._onAnswer(false)}
        />
      </View>
    </View>
    )
  };

  resultsScreen = () => {
    const { deck, navigation } = this.props;
    const { currentQuestionIndex, correctAnswers } = this.state;
    if (currentQuestionIndex === deck.questions.length) {
      return (
        <View style={styles.container}>
          <View style={styles.textContainer}>
            <Text style={styles.resultTextStyle}>
              You have completed the quiz!
            </Text>
            <Text style={styles.resultTextStyle}>
              You scored {correctAnswers}/{deck.questions.length} questions correctly!
            </Text>
          </View>
          <View style={styles.inputContainer}>
            <Button
              style={styles.btnStyle}
              label="Retry Quiz"
              onPress={this.resetQuiz}
            />
            <Button
              style={styles.btnStyle}
              label="Back to Deck"
              onPress={() => navigation.goBack()}
            />
          </View>
        </View>
      )
    }
  };

  render() {
    const { deck } = this.props;

    // No questions in deck
    if (!deck || !deck.questions || !deck.questions.length === 0) {
      return (
        <View style={styles.container}>
          <Text>Sorry there are no questions in this deck.</Text>
        </View>
      );
    }

    // Show questions in the deck.
    return (
      <View style={styles.container}>
        { deck.questions.map((question, index) => this.renderQuestion(question, index)) }
        { this.resultsScreen() }
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-start',
  },
  questionNumber: {
    fontWeight: 'bold',
    alignSelf: 'flex-start',
    padding: 10
  },
  textContainer: {
    flex: 1,
    justifyContent: 'center',
    alignSelf: 'center',
    padding: 10
  },
  inputContainer: {
    flex: 1,
    justifyContent: 'center',
    padding: 10
  },
  btnStyle: {
    marginTop: 10,
    marginBottom: 10,
    marginLeft: 40,
    marginRight: 40,
  },
  textStyle: {
    color: black,
    fontSize: 36,
    alignSelf: 'center',
    paddingTop: 20,
    paddingBottom: 20
  },
  textBtnStyle: {
    color: blue,
    fontSize: 24,
    alignSelf: 'center',
    paddingTop: 20,
    paddingBottom: 40,
  },
  resultTextStyle: {
    color: black,
    fontSize: 16,
    alignSelf: 'center',
    paddingTop: 20,
    paddingBottom: 20
  },
});

Quiz.propTypes = {
  deck: PropTypes.object.isRequired,
  navigation: PropTypes.object.isRequired,
  fetchDeck: PropTypes.func.isRequired,
};

Quiz.defaultPropTypes = {
  deck: {}
};

const mapStateToProps = (state, ownProps) => {
  const { navigation } = ownProps;
  const title = get(navigation, 'state.params.title', '');
  const decksList = state.decks;
  return {
    deck: title ? decksList[title] : {}
  };
};

export default connect(mapStateToProps, {
  fetchDeck
})(Quiz);