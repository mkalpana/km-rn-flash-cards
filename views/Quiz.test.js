import React from 'react';
import { Quiz } from './Quiz';
import renderer from 'react-test-renderer';

const deck = { title: 'React', questions: [] };

describe('Quiz', () => {
  it('renders without crashing', () => {
    const rendered = renderer.create(<Quiz
      deck={deck}
      fetchDeck={() => deck}
      navigation={{
        goBack: null,
        state: { params: { title: 'React' } }
      }}

    />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
