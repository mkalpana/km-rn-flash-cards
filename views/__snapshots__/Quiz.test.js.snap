// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz renders without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "center",
          "flex": 1,
          "justifyContent": "center",
          "padding": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "fontSize": 16,
            "paddingBottom": 20,
            "paddingTop": 20,
          }
        }
      >
        You have completed the quiz!
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "fontSize": 16,
            "paddingBottom": 20,
            "paddingTop": 20,
          }
        }
      >
        You scored 
        0
        /
        0
         questions correctly!
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "padding": 10,
        }
      }
    >
      <View
        style={
          Object {
            "marginBottom": 10,
            "marginLeft": 40,
            "marginRight": 40,
            "marginTop": 10,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "black",
                  "borderColor": "black",
                  "borderRadius": 5,
                  "borderWidth": 2,
                  "padding": 20,
                  "paddingLeft": 40,
                  "paddingRight": 40,
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "textAlign": "center",
                  },
                  undefined,
                ]
              }
            >
              Retry Quiz
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "marginBottom": 10,
            "marginLeft": 40,
            "marginRight": 40,
            "marginTop": 10,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "black",
                  "borderColor": "black",
                  "borderRadius": 5,
                  "borderWidth": 2,
                  "padding": 20,
                  "paddingLeft": 40,
                  "paddingRight": 40,
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "textAlign": "center",
                  },
                  undefined,
                ]
              }
            >
              Back to Deck
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
